### PR TITLE
Feature/datasource list in api editor

### LINKED ISSUE #4567: Show a list of available datasources in API editor

### PR DESCRIPTION
**Pull request by arunvjn**
## Description
API pane improvements

- Showing a list of datasources to allow users to navigate to it and make changes
- Enhance datasource autocomplete in API pane 

Fixes #4567 

## Type of change
- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feature/datasource_list_in_api_editor 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 51.55 **(-0.06)** | 32.53 **(0.08)** | 29.63 **(-0.19)** | 52.05 **(-0.04)**
 :red_circle: | app/client/src/components/ads/Tabs.tsx | 27.5 **(1.31)** | 40 **(-4.44)** | 0 **(0)** | 27.5 **(1.31)**
 :red_circle: | app/client/src/components/ads/Text.tsx | 100 **(0)** | 75 **(-1.92)** | 100 **(0)** | 100 **(0)**
 :green_circle: | app/client/src/components/editorComponents/form/fields/EmbeddedDatasourcePathField.tsx | 29.37 **(4.37)** | 2.7 **(0)** | 4 **(-0.35)** | 31.3 **(4.63)**
 :sparkles: :new: | **app/client/src/pages/Applications/ProductUpdatesModal/ReleaseComponent.tsx** | **22.67** | **47.06** | **2.7** | **24.64**
 :sparkles: :new: | **app/client/src/pages/Editor/APIEditor/DatasourceList.tsx** | **52.78** | **46.15** | **0** | **56.25**
 :green_circle: | app/client/src/pages/Editor/APIEditor/Form.tsx | 37.5 **(-0.17)** | 46.81 **(2.62)** | 0 **(0)** | 39.47 **(0.18)**
 :green_circle: | app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx | 28.57 **(0.08)** | 23.4 **(0)** | 3.33 **(0.1)** | 32.03 **(0.78)**</details>

### CODE REVIEW

**Comment by hetunandu** [comment]
Deleting a rest datasource takes me to query datasource screen

recording


<details><summary>Browser metadata</summary>





```
Path:      /applications/6034d9f5438ec90ecf230a5c/pages/6034d9f5438ec90ecf230a5e/edit/api/6034da0c438ec90ecf230a5f
Browser:   Chrome 90.0.4430.212 on Mac OS 10.15.7
Viewport:  1680 x 939 @2x
Language:  en-GB
Cookies:   Enabled
```

[View on BrowserStack](https://live.browserstack.com/dashboard#url=https%3A%2F%2Fdeploy-preview-4587--appsmith-editor-ce.netlify.app%2Fapplications%2F6034d9f5438ec90ecf230a5c%2Fpages%2F6034d9f5438ec90ecf230a5e%2Fedit%2Fapi%2F6034da0c438ec90ecf230a5f&os=OS+X&os_version=Catalina&browser=Chrome&browser_version=90.0&resolution=1680x939&speed=1&start=true)



</details>

[Open Deploy Preview](https://deploy-preview-4587--appsmith-editor-ce.netlify.app/applications/6034d9f5438ec90ecf230a5c/pages/6034d9f5438ec90ecf230a5e/edit/api/6034da0c438ec90ecf230a5f) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=60a607b08dfeda0007f920ab&commentID=60a61d5490bb9ce1e5e2d2e0&resolution=resolved)

**Comment by hetunandu** [comment]
Clicking on the datasource does not apply it on the api instead navigates me to that datasource

recording


<details><summary>Browser metadata</summary>





```
Path:      /applications/6034d9f5438ec90ecf230a5c/pages/6034d9f5438ec90ecf230a5e/edit/queries?importTo=6034d9f5438ec90ecf230a5e
Browser:   Chrome 90.0.4430.212 on Mac OS 10.15.7
Viewport:  1680 x 939 @2x
Language:  en-GB
Cookies:   Enabled
```

[View on BrowserStack](https://live.browserstack.com/dashboard#url=https%3A%2F%2Fdeploy-preview-4587--appsmith-editor-ce.netlify.app%2Fapplications%2F6034d9f5438ec90ecf230a5c%2Fpages%2F6034d9f5438ec90ecf230a5e%2Fedit%2Fqueries%3FimportTo%3D6034d9f5438ec90ecf230a5e&os=OS+X&os_version=Catalina&browser=Chrome&browser_version=90.0&resolution=1680x939&speed=1&start=true)



</details>

[Open Deploy Preview](https://deploy-preview-4587--appsmith-editor-ce.netlify.app/applications/6034d9f5438ec90ecf230a5c/pages/6034d9f5438ec90ecf230a5e/edit/queries?importTo=6034d9f5438ec90ecf230a5e) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=60a607b08dfeda0007f920ab&commentID=60a61dc24e1e4e0301ecafca&resolution=resolved)

**Comment by arunvjn** [comment]
> Clicking on the datasource does not apply it on the api instead navigates me to that datasource
> 
> recording
> 
> Browser metadata
> [Open Deploy Preview](https://deploy-preview-4587--appsmith-editor-ce.netlify.app/applications/6034d9f5438ec90ecf230a5c/pages/6034d9f5438ec90ecf230a5e/edit/queries?importTo=6034d9f5438ec90ecf230a5e) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=60a607b08dfeda0007f920ab&commentID=60a61dc24e1e4e0301ecafca&resolution=resolved)

This is the expected behaviour. The users will now be able to see a list of datasource without having to use them.
This would also let them delete datasource.

File: app/client/src/pages/Editor/APIEditor/Form.tsx [hunk 1]
```diff
@@ -376,173 +550,206 @@ function ApiEditorForm(props: Props) {
...
-                        <ShowHideImportedHeaders
-                          onClick={(e) => {
-                            e.preventDefault();
-                            toggleInheritedAttributes(!showInheritedAttributes);
-                          }}
-                        >
-                          <ButtonIcon
-                            icon={
-                              showInheritedAttributes ? "eye-open" : "eye-off"
+      {hintMessages && (
```
Comment (hetunandu) [inline app/client/src/pages/Editor/APIEditor/Form.tsx:553]: This component is getting too huge. Lets split it up into smaller components

File: app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx [hunk 2]
```diff
@@ -307,20 +307,10 @@ class DatasourceRestAPIEditor extends React.Component<Props> {
...
-    } = this.props;
+    const { isNewDatasource, isSaving, pluginImage } = this.props;
     return (
       <>
-        <BackButton
-          onClick={() =>
-            history.push(DATA_SOURCES_EDITOR_URL(applicationId, pageId))
-          }
-        />
+        <BackButton onClick={history.goBack} />
```
Comment (hetunandu) [inline app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx:313]: If I come here after deleting a datasource, clicking back will take me to the deleted datasource?

**Review by hetunandu — CHANGES_REQUESTED** [review]
Needs refactor

File: app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx (hunk 2 above)
Comment (arunvjn) [reply app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx:313]: This is the back button you see datasource form page. Clicking the back button will take you back to which ever screen you were on. The use case of taking the user back to the previous page on datasource delete is not handled here. @hetunandu

File: app/client/src/pages/Editor/APIEditor/Form.tsx (hunk 1 above)
Comment (arunvjn) [reply app/client/src/pages/Editor/APIEditor/Form.tsx:661]: Separated out imported headers from headers component.

**Comment by arunvjn** [comment]
> Deleting a rest datasource takes me to query datasource screen
> 
> recording
> 
> Browser metadata
> [Open Deploy Preview](https://deploy-preview-4587--appsmith-editor-ce.netlify.app/applications/6034d9f5438ec90ecf230a5c/pages/6034d9f5438ec90ecf230a5e/edit/api/6034da0c438ec90ecf230a5f) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=60a607b08dfeda0007f920ab&commentID=60a61d5490bb9ce1e5e2d2e0&resolution=resolved)

I haven't modified this behaviour. Shall we take this up in a separate issue ?

**Review by hetunandu — APPROVED** [review]

**Pull request merged**